### ISSUE TITLE
Implement AtomMapping container

### ISSUE THREAD

**Issue by richardjgowers** [OpenFreeEnergy/openfe#6]
Should be a read only simple data class holding enough information for mapping two sets of ints (atom indices) to each other...

```python
class AtomMapping:
    AtoB : dict[int, int]
    BtoA : dict[int, int]
```

**Comment by dwhswenson** [comment]
Would it make sense for the `AtomMapping` container to also carry a `context` (or `cache`) dict (default empty or None)? Here's my thought in how that could be useful for when implementing practical `AtomMapper`s (also relates to #5, #9, #4):


```python
# wrapping stuff from lomap.mcs
class MCSAtomMapper(AtomMapper):
    def __init__(self, **lomap_kwargs):
        self.lomap_kwargs = lomap_kwargsA

    def _context(self, mol1, mol2):
        return {'lomap': lomap.MCS(mol1, mol2, **self.lomap_kwargs)}

    def _propose_from_context(self, mol1, mol2, context):
        lomap_mcs = context['lomap']
        return [AtomMapping(get_lomap_mapping(mol1, mol2, lomap_mcs),  # some magic 🤷‍♂️ 
                            get_lomap_mapping(mol2, mol1, lomap_mcs),
                            context=context})]

    def propose_mappings(self, mol1, mol2):
        context = self._context(mol1, mol2)
        for proposal in self._propose_from_context(mol1, mol2, context):
            yield proposal


# this is a lazy way to implement it, but just supposed to be a skeleton
class MCSScore(Scorer):
    def __init__(self, func_name):
        self.func_name = func_name

    def score(self, atommapping):
        try:
            lomap_mcs = atommapping.context['lomap']
        except KeyError:
            raise TypeError("MCSScore must use an atommapping from an "
                            "MSCAtomMapper")
        func = getattr(lomap_mcs, self.func_name)
        return func()
```

Usage would then look something like:

```python
mapper = MCSAtomMapper(...)
scorer = MCSScore('heterocycles_rule')
for mapping in mapper.propose_mappings(mol1, mol2):
    score = scorer(mapping)
    ... # do something with score
```

**Comment by richardjgowers** [comment]
@dwhswenson  Ah right, I just saw that the context goes into the mapping.  Not a fan, it makes the object all intertwined.  I'd sooner have a cache in the `AtomMapper` and have it serve as both mapper and scorer. 

```python
class MagicAtomMapper:
    def __init__(self):
        ...
        self._cache = {}

    def _generate_mappings(self, m1, m2):
        # actual mapping work here

    def propose_mappings(self, m1, m2):
        for mapping in self._generate_mappings(m1, m2):
            self._cache[mapping] = score  # assume that we calculate score as we're mapping, so save it for later
            yield mapping

    def score(self, mapping):
        try:
            return self._cache[mapping]
        except KeyError:
            something clever?


mapper = MagicAtomMapper()
# mapper object can also quack like a scorer?
scorer = mapper
```

**Comment by dwhswenson** [comment]
Fair enough. But maybe we make `Scorer.score` and `Scorer.annotation` private? These are conveniences for implementers; I think the API should only require that `Scorer.__call__` takes a mapping and returns a `(score, annotation)` namedtuple.

In that case, `scorer = mapping.score`. I prefer this to making mappers duck-type like scorers. It seems counterintuitive to me that `__call__` for a mapper would produce a score (I would expect a mapping.)

**Comment by dwhswenson** [comment]
Also, in the specific context of MCS, which has several scoring functions associated with one mapper, users could either do:

```python
# may auto-update if other scorers are implemented within the lomap class?
scorer = mcs_mapper.get_scorer('heterocycles_rule')
```

or

```python
# predefine the functions we support
scorer = mcs_mapper.heterocycles_scorer
```

**Comment by dwhswenson** [comment]
Moving to done: the functionality targeted here was implemented in #28 (with some updates/improvements from the original proposal).
